# SAMPLING

### DEFINITION:

Sampling is the process of selecting a smaller, representative set of data points from a large dataset.

Sampling helps to;
1. reduce computation time, and training time.
2. reduce memory/storage requirements.

#### EXAMPLE:
Suppose a list has 500 items of groceries, and to check randomly at counter to weather all the items are packed, a sample of 25 groceries were collected from list.

In [21]:
import random
groceries = list(range(1,500))
sample = random.sample(groceries, 111)
print(f" GROCERIES=: {sample [:12]}" )

 GROCERIES=: [446, 56, 100, 116, 331, 148, 21, 312, 80, 15, 74, 260]


A sample of 111 groceries are selected from list of 500 groceries.

## RANDOM SAMPLING:

### DEFINITION:

Random sampling is a technique in which data points are selected randomly from a larger dataset, so that each data point has an equal or known chance of being selected.

#### Example:
Suppose we have 20 numbers;

In [22]:
import random
numbers = [3, 7, 13, 17, 23, 27, 33, 37, 43, 47, 53, 57, 63, 67, 73, 77, 83, 87, 93, 97]
sample = random.sample(numbers, 7)
print("Random Sample: ", sample)

Random Sample:  [23, 47, 63, 43, 13, 33, 53]


Here, random.sample() randomly selects 7 numbers from the original list.
20 numbers → randomly select 7 numbers  → sample data

## STRATIFIED SAMPLING:

### DEFINITION:
Dividing the data into groups first, then randomly selecting data from every group.

Stratified Sampling helps to;
1. Easy to implement
2. When properly performed, every record has an equal opportunity to be selected

##### Real-time Examples:

1. Banking: Select customers from different income groups.
2. Education: Select students from different departments or years.

#### Example:

An Villa has differnt blocks as Apartments
*  BLOCK-A 
*  BLOCK-B
*  BLOCK-C
*  BLOCK-D

Owner wants to select 3 houses as winners fromm each BLOCK.

In [29]:
Block_1 = list(range(101, 110))
Block_2 = list(range(201, 210))
Block_3 = list(range(301, 310))
Block_4 = list(range(401, 410))

import random

sample = (
    random.sample(Block_1, 3) +
    random.sample(Block_2, 3) +
    random.sample(Block_3, 3) +
    random.sample(Block_4, 3)
)

print(sample[:12])

[107, 102, 101, 205, 204, 209, 306, 303, 309, 401, 409, 405]


Every BLOCK is represented equally in the sample.

## CLUSTER SAMPLING:

### DEFINITION:

* Cluster sampling is a sampling technique where a large population is divided into groups called clusters, and then some entire clusters are randomly selected for the study.

* Instead of randomly selecting individual data points, we randomly select groups of data points.

##### Real-time Examples:
* Marketing: Select cities and analyze customers from those cities.

* Government surveys: Select villages or districts and collect information from households.

#### EXAMPLE:
Imagine a company has  Employees data from 4 Companies:

Department A → 100 Employees
Department B → 800 Employees
Department C → 1200 Employees
Department D → 900 Employees

Instead of randomly selecting individual Employees from all Departments, the company randomly selects 2 Departments and analyzes all Employees from those Departments.

This is cluster sampling because the DEPARTMENTS are the clusters.

In [41]:
import random

stores = {
    "Department_A": ["Employee1", "Employee2", "Employee3", "Employee4", "Employee5", "Employee6"],
    "Department_B": ["Employee7", "Employee8", "Employee9", "Employee10"],
    "Department_C": ["Employee1", "Employee2", "Employee3", "Employee4", "Employee5", "Employee6", "Employee7"],
    "Department_D": ["Employee11", "Employee12", "Employee13", "Employee14", "Employee15"]
}

# Randomly select 2 clusters (stores)
selected_Department = random.sample(list(stores.keys()), 2)

# Collect all data from selected clusters
sample = []

for store in selected_Department:
    sample.extend(stores[store])

print("Selected Department:", selected_Department)
print("Sample Data:", sample)


Selected Department: ['Department_D', 'Department_A']
Sample Data: ['Employee11', 'Employee12', 'Employee13', 'Employee14', 'Employee15', 'Employee1', 'Employee2', 'Employee3', 'Employee4', 'Employee5', 'Employee6']


##### Code Explanation:

Here:

Department A, B, C, D → Clusters

Store A and D → Randomly selected clusters

All Employees from A and D → Sample

Cluster sampling is a sampling technique in which the population is divided into groups called clusters, and some clusters are randomly selected to form the sample.

## SYSTEMATIC SAMPLING:

### DEFINITION:
Systematic sampling is a sampling technique where we select data points at a fixed regular interval from a dataset after choosing a starting point randomly.

##### Real-time Examples:
* Manufacturing: Inspect every 20th product from a production line.

* Education: Select every 10th student from a student list.

* Healthcare: Select every 5th patient record for analysis.

#### Example:
Suppose we have 88 online orders and want to select every 8th order.

In [47]:
orders = [f"Order_{i}" for i in range(1, 88)]

# Select every n3rd order
sample = orders[2::3]

print(f"Systematic Sample: , {sample}")

Systematic Sample: , ['Order_3', 'Order_6', 'Order_9', 'Order_12', 'Order_15', 'Order_18', 'Order_21', 'Order_24', 'Order_27', 'Order_30', 'Order_33', 'Order_36', 'Order_39', 'Order_42', 'Order_45', 'Order_48', 'Order_51', 'Order_54', 'Order_57', 'Order_60', 'Order_63', 'Order_66', 'Order_69', 'Order_72', 'Order_75', 'Order_78', 'Order_81', 'Order_84', 'Order_87']


###### Code Explanation:

The starting point can change, but after that, we consistently select every 4th record.

## SAMPLING BIAS:

### DEFINITION:

Sampling bias happens when the way we choose data causes the sample to be unfair or unrepresentative of the real population.

##### Real-time Examples:

* Banking: Include customers from different income groups.

* Marketing: Include customers with different ages, locations, and preferences.

* AI/ML: Build training datasets that represent the real-world population.

#### Example:
The Comany CEO surveys only  Department_C Employees about the Population.

The opinions of Department_A, Department_B and Department_D Employees are ignored.

In [56]:
Department_C_Employees = ["Employee1", "Employee2", "Employee3", "Employee4", "Employee5", "Employee6", "Employee7"]
print(Department_C_Employees[2::2])

['Employee3', 'Employee5', 'Employee7']


##### Code Explanation:
The Owner wants to analyse only odd indexed Employees from Department_C_Employees

## SAMPLING ERROR:

### DEFINITION:

Sampling error is the difference between the result obtained from a sample and the actual result of the entire population, simply because we studied only a subset of the data.

In simple words: Sampling error happens because a sample may not give exactly the same result as the complete population.

Important: Sampling error is usually due to random variation, not because someone selected the data unfairly. This is different from sampling bias.

##### Real-time Examples:
* Education: Estimate average student marks from a sample of students.

* Polling: Estimate public opinion by surveying a sample of voters.

* AI/ML: Evaluate how well a model performs when only a subset of available data is used.

#### EXAMPLE:

Suppose:

Product average Rating = 4.8, 
Sample average Rating = 5
The difference (0.2) is the sampling error.

In [58]:
product_rating = 4.8
sample_rating = 5.0

sampling_error = sample_rating - product_rating

print(f" Sampling_Error is: {sampling_error}" )

 Sampling_Error is: 0.20000000000000018


##### Code Explanation:
The Difference between Original Product Rating and Collected Sample Rating is defeeined as, Sampling Error. i.e; 0.20000000000000018 for the above example.

## SAMPLING SIZE:

### DEFINITION:

Sample size is the number of data points, observations, or individuals selected from a population for a study or machine-learning task.

In simple words: Sample size tells us how many data points we are going to take from the entire dataset.

##### Real-time Examples:

* Banking: Analyze a sample of transactions for fraud detection.

* Manufacturing: Inspect a certain number of products for quality control.

#### EXAMPLE:
Suppose we have 200000 Employee records in a Company, and CEO wants to select a sample of 121 Employees.

In [62]:
import random

employees = list(range(1, 20001))

sample_size = 121

sample = random.sample(employees, sample_size)

print("Sample Size:", len(sample))
print("Selected Employees:", sample)


Sample Size: 121
Selected Employees: [18611, 303, 13783, 10107, 15818, 10464, 15654, 12678, 7134, 573, 14557, 2576, 1172, 3040, 15209, 865, 7580, 16068, 746, 7707, 17265, 18864, 2398, 16436, 3875, 5014, 2800, 4688, 14148, 11599, 3107, 10458, 16574, 13877, 1601, 8742, 10503, 4773, 18008, 13487, 9638, 14749, 8420, 19481, 11554, 972, 9233, 2174, 836, 2146, 9389, 5214, 5901, 12247, 8580, 18542, 18412, 15263, 2027, 2393, 14191, 12214, 8463, 562, 11210, 8027, 19867, 10588, 2577, 18492, 8983, 8102, 15152, 19455, 14514, 8033, 5899, 13270, 17907, 288, 19246, 9200, 19844, 3103, 18012, 7551, 5006, 1573, 16210, 14965, 1590, 17483, 12545, 10168, 13039, 16415, 17166, 2156, 19442, 10599, 14773, 12359, 3738, 19597, 5197, 4905, 6617, 2859, 5200, 16407, 7253, 63, 2841, 3458, 16188, 128, 13482, 6781, 14778, 11184, 10400]


##### Code Explanation:

This can make experimentation faster and reduce computational requirements.

However, more data is not automatically better if the additional data is poor quality, redundant, or unrepresentative.